### PR TITLE
(309) Update service to handle LAOs in a more user friendly way

### PR DESCRIPTION
**Pull request by richpjames**
# Context
There are certain people in the prison system who are high profile and therefore the information about them is restricted for their protection, these people are referred to as LAOs (limited access offenders). Some people may be a regular PiP for some senior staff but an LAO for other staff.

This API PR means that if a LAO is returned from the API the UI will now see only their CRN: https://github.com/ministryofjustice/hmpps-approved-premises-api/pull/804. 
Previously this would have thrown a non-user friendly error.
[Trello](https://trello.com/c/IdrDyeuM/309-cannot-list-bookings-from-premises-with-lao-offenders-%F0%9F%90%9E)

As there are many places in the application where the person object is used this means we have had to make changes to all of these places. More often than not we simply use a `Person`s name which we can simply fill in 'the person' instead but in some places it is more complicated than that.

# Changes in this PR
- We update to the appropriate types from the API branch
- We update the factories that use the Person object to return either a RestrictedPerson or a FullPerson
- We add some utils for dealing with the two new types of person
- We adjust the People Controller to throw an error if a restricted person is returned from the API
- We update the Apply/Booking creation journeys to deal with the new error thrown from the People Controller
- We update the various dashboards to handle restricted persons
- We update the Apply journey to handle restricted persons
- We update the adminIdentityBar

## Screenshots of UI changes
Apply -- throws an error if the the CRN entered is an LAO

### CODE REVIEW

File: integration_tests/support/index.ts [hunk 1]
```diff
@@ -2,5 +2,5 @@ import './commands'
 import 'cypress-axe'
 
 Cypress.Keyboard.defaults({
-  keystrokeDelay: 1,
+  keystrokeDelay: 0,
```
Comment (bryangaledxw) [inline integration_tests/support/index.ts:5]: 👀

File: server/utils/assessments/tableUtils.ts [hunk 2]
```diff
@@ -24,18 +25,24 @@ const getStatus = (assessment: AssessmentSummary): string => {
...
+      text: linkText || person.name,
       hiddenText,
       attributes: { 'data-cy-assessmentId': assessment.id, 'data-cy-applicationId': assessment.applicationId },
     },
   )
 }
 
+export const restrictedPersonCell = (person: RestrictedPerson) => {
+  return {
+    text: `${person.crn} is a LAO`,
```
Comment (bryangaledxw) [inline server/utils/assessments/tableUtils.ts:42]: Should this be consistent with the booking table? So just "Limited access offender"? Or "LAO CRN: ${person.crn}", so it's consistent with applications?

File: server/controllers/peopleController.ts [hunk 3]
```diff
@@ -14,8 +16,14 @@ export default class PeopleController {
...
           const person = await this.personService.findByCrn(req.user.token, crn, !!checkCaseload)
           req.flash('crn', person.crn)
+
+          if (!isFullPerson(person)) {
+            throw new RestrictedPersonError(person.crn)
+          }
         } catch (err) {
-          if ('data' in err && err.status === 404) {
+          if (err.type === 'RESTRICTED_PERSON') {
+            addErrorMessageToFlash(req, err.message, 'crn')
```
[the remark this answers is not in the document]
Comment (bryangaledxw) [reply server/controllers/peopleController.ts:25]: Ah, I think I understand - whether we get a full person or a restricted person depends on who the user is. So one person might be able to create an application for a given CRN because they have access to the full record, but another person might just see it in the dashboard and shouldn't be able to create an application for the person

File: server/utils/errors.ts [hunk 4]
```diff
@@ -31,3 +31,12 @@ export class ErrorWithData extends Error {
     super()
   }
 }
+
+export class RestrictedPersonError extends Error {
```
Comment (bryangaledxw) [inline server/utils/errors.ts:35]: Possibly move this into its own commit, and move to before its first use

File: server/controllers/peopleController.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (richpjames) [reply server/controllers/peopleController.ts:25]: thanks, I'll see if I can make it a bit clearer

File: server/utils/applications/utils.test.ts [hunk 5]
```diff
@@ -149,17 +152,19 @@ describe('utils', () => {
   describe('dashboardTableRows', () => {
     it('returns an array of applications as table rows', async () => {
       ;(tierBadge as jest.Mock).mockReturnValue('TIER_BADGE')
-      const arrivalDate = DateFormats.dateObjToIsoDate(new Date(2021, 0, 3))
+      ;(isFullPerson as unknown as jest.Mock).mockReturnValue(true)
```
[the remark this answers is not in the document]
Comment (richpjames) [reply server/utils/applications/utils.test.ts:155]: Yeah this is better, we should probably use `MockedFunction` more widely
